modules/gapi/test/common/gapi_core_tests_inl.hpp

```diff
@@ -1377,6 +1386,171 @@ TEST_P(NormalizeTest, Test)
+    {
+        EXPECT_TRUE(compact_gapi == compact_ocv);
+        EXPECT_TRUE(compareVectorsAbsExact(labels_gapi, labels_ocv));
+        EXPECT_TRUE(compareVectorsAbsExact(centers_gapi, centers_ocv));
+    }
+}
+
+TEST_P(KMeans3DNoInitTest, AccuracyTest)
+{
+    const int quantity = sz.height;
+    const cv::TermCriteria criteria(TermCriteria::MAX_ITER + TermCriteria::EPS, 30, 0);
+    const int attempts = 1;
+    double compact_gapi = -1.;
+    std::vector<cv::Point3f> in_vector{}, centers_gapi{};
+    std::vector<int> labels_gapi{};
+    initPointsVectorRandU(quantity, in_vector);
+    // G-API code //////////////////////////////////////////////////////////////
+    cv::GArray<cv::Point3f> in, centers;
+    cv::GArray<int> labels, gLabelsIn(std::vector<int>{});
+    cv::GOpaque<double> compactness;
+    std::tie(compactness, labels, centers) =
+        cv::gapi::kmeans(in, K, gLabelsIn, criteria, attempts, flags);
+    cv::GComputation c(cv::GIn(in), cv::GOut(compactness, labels, centers));
+    c.apply(cv::gin(in_vector), cv::gout(compact_gapi, labels_gapi, centers_gapi), getCompileArgs());
+    // Validation //////////////////////////////////////////////////////////////
+    {
+        EXPECT_GE(compact_gapi, 0.);
+        EXPECT_EQ(labels_gapi.size(), static_cast<size_t>(quantity));
+        EXPECT_EQ(centers_gapi.size(), static_cast<size_t>(K));
+    }
+}
+
+TEST_P(KMeans3DInitTest, AccuracyTest)
+{
+    const int quantity = sz.height;
+    const cv::TermCriteria criteria(TermCriteria::MAX_ITER + TermCriteria::EPS, 30, 0);
+    const int attempts = 1;
+    double compact_ocv = -1., compact_gapi = -1.;
+    std::vector<int> in_labels(quantity);
+    std::vector<int> labels_ocv{}, labels_gapi{};
+    std::vector<cv::Point3f> in_vector{}, centers_ocv{}, centers_gapi{};
+    initPointsVectorRandU(quantity, in_vector);
+    cv::randu(in_labels, 0, K);
+    labels_ocv = in_labels;
+    // G-API code //////////////////////////////////////////////////////////////
+    cv::GArray<cv::Point3f> in, centers;
+    cv::GArray<int> gLabelsIn, gLabelsOut;
+    cv::GOpaque<double> compactness;
+    std::tie(compactness, gLabelsOut, centers) =
+        cv::gapi::kmeans(in, K, gLabelsIn, criteria, attempts, flags);
+    cv::GComputation c(cv::GIn(in, gLabelsIn), cv::GOut(compactness, gLabelsOut, centers));
+    c.apply(cv::gin(in_vector, in_labels), cv::gout(compact_gapi, labels_gapi, centers_gapi),
+            getCompileArgs());
+    // OpenCV code /////////////////////////////////////////////////////////////
+    compact_ocv = cv::kmeans(in_vector, K, labels_ocv, criteria, attempts, flags, centers_ocv);
+    // Comparison //////////////////////////////////////////////////////////////
+    {
+        EXPECT_TRUE(compact_gapi == compact_ocv);
+        EXPECT_TRUE(compareVectorsAbsExact(labels_gapi, labels_ocv));
+        EXPECT_TRUE(compareVectorsAbsExact(centers_gapi, centers_ocv));
+    }
```

Comment:
Where do you test the case when you try to call ``kmeans`` multiple times with the different ``initLabels``?

Reply:
Did you mean multiple calls of `kmeans` with `attempts=1` and manual setup of `initLabels` every time? This case wouldn't differ much from the case with a single `kmeans` call.
You could mean calling `kmeans` with `attempts != 1` though. If so, such a case can't be really tested, because after the first attempt, `kmeans` initializes `initLabels` randomly and thus we cannot compare the output with the corresponding `cv::kmeans`